### ISSUE TITLE
typescript: do we need a module definition in index.d.ts

### ISSUE THREAD

**Issue by ggascoigne** [tannerlinsley/react-table#1580]
I'm finding that react-table exports only seem to work if I wrap all of index.d.ts in a `declare module 'react-table'` block.

I'd assumed that being exposed in the npm package would give it a default 'react-table' module, but that doesn't seem to be the case.

@stramel, am I missing something?

**Comment by stramel** [comment]
@ggascoigne I'm not sure what you mean? We aren't currently exposing the types in this package.

**Comment by ggascoigne** [comment]
from the docs:

```txt
Also note that if your main declaration file is named index.d.ts and lives at the root of the package (next to index.js) you do not need to mark the "types" property, though it is advisable to do so.
```

So yeah, they are exposed.

**Comment by ggascoigne** [comment]
this is what was masing a couple of the optional types that I kept seeing that you weren't - which incidentally are now fixed in the last update.

**Comment by stramel** [comment]
Hmmm, I assumed that since our file was at `dist/index.js` it wouldn't be. I would vote we move it to `types/index.d.ts` then so it isn't exposed by default.

**Comment by ggascoigne** [comment]
Or just wrap it in a module declaration and let people use it and report bugs.

It really is quite good enough to use for a lot of use cases.

**Comment by ggascoigne** [comment]
as an aside, I read the docs the way that you did as well, until I found that deleting the old types actually changed things.

**Comment by stramel** [comment]
@ggascoigne Hmmmm removed due to a recommendation from dtslint. 

https://github.com/microsoft/dtslint/blob/master/docs/no-declare-current-package.md

/cc @ChristianMurphy

**Comment by ChristianMurphy** [comment]
@ggascoigne I'm not sure I'm following the issue, could you provide an example?

`import * as ReactTable from 'react-table'` would import the full namespace if needed.
`import { UseSortByColumnOptions } from 'react-table'` can be used to get at optional types.

**Comment by ggascoigne** [comment]
OK, turns out it's a bit more complex than I was hoping.

If I don't have a local file that extends the types, then yes, the types do appear to work correctly, but as soon as I add a file with an additional declaration, say a file with:

```tsx
declare module 'react-table' {
  export interface TableOptions<D extends object>  // UseExpandedOptions<D>,
    extends UseFiltersOptions<D>,
      // UseGroupByOptions<D>,
      UsePaginationOptions<D>,
      UseRowSelectOptions<D>,
      UseSortByOptions<D>,
      UseFiltersOptions<D>,
      UseResizeColumnsOptions<D>{}
...
```

Then the types that were declared in the library are no longer visible.

If both start with `declare module` then the declaration merging works as desired, if neither contain the declaration then the declaration merging doesn't seem to work.  It's as though the types from the library are in a module that resolves correctly, but it has some different name that isn't `react-table`.

**Comment by ChristianMurphy** [comment]
Does adding import syntax to the top of the override typing, allow the types used for extension to be brought in?

**Comment by ggascoigne** [comment]
I think that that was it, thanks :)

I'll post an update to the `react-table-config.d.ts` file.

**Comment by ggascoigne** [comment]
https://github.com/tannerlinsley/react-table/pull/1582

**Comment by ChristianMurphy** [comment]
@ggascoigne was this resolved with #1582 ?

**Comment by ggascoigne** [comment]
Yes. Thanks for the reminder.
